### ISSUE TITLE
Implement throttling

### ISSUE THREAD

**Issue by dbrgn** [dbrgn/sekursranko#2]
User configurable, e.g. based on IP (this is only feasible when serving a small user base due to things like CGNAT).

**Comment by TimWolla** [comment]
> Sekurŝranko is meant to be run behind a reverse proxy (e.g. Nginx) that does TLS termination. That's why it currently doesn't support TLS directly.

It probably makes sense to leave rate limiting to the reverse proxy, which has a better overview over the whole system state, then.

**Comment by dbrgn** [comment]
Yes, I am currently unsure whether this library should implement throttling at all.

A proper implementation would have to support multiple processes in parallel, this is something that would probably require an additional dependency like Redis. It's probably better to leave that to the reverse proxy (e.g. [ngx_http_limit_req](http://nginx.org/en/docs/http/ngx_http_limit_req_module.html)).

On the other hand we could simply add a naïve implementation based on IP that will work for the single-user or few-users use case.

**Comment by TimWolla** [comment]
> On the other hand we could simply add a naïve implementation based on IP that will work for the single-user or few-users use case.

You would have to ingest the IP from the reverse proxy then (proxy_protocol, X-Forwarded-For, etc.), which adds to the complexity.

**Comment by rugk** [comment]
Why not implement it based on the `threemaSafeBackupId`? That is likely easier, more resistant (as attackers can easily change IP addresses) and I guess also easier to implement. (proxy is irrelevant then)
If you want to add IP throttling you can then anyway do it at nginx or so, for the whole domain/proxy…

Described on https://threema.ch/press-files/2_documentation/cryptography_whitepaper.pdf page 21.

**Comment by dbrgn** [comment]
> Why not implement it based on the threemaSafeBackupId?

This is not about uploading data, but about downloading / brute forcing backups in bulk :slightly_smiling_face: Using the backup id is useless in that case.

(Note that even if you manage to download an encrypted backup, you still neither know who uploaded it nor how to decrypt it. It's mostly about defense in depth, for single-user use cases.)

> You would have to ingest the IP from the reverse proxy then (proxy_protocol, X-Forwarded-For, etc.)

Yup (or alternatively the [Forwarded](https://tools.ietf.org/html/rfc7239) header).

**Comment by rugk** [comment]
> This is not about uploading data, but about downloading / brute forcing backups in bulk slightly_smiling_face Using the backup id is useless in that case.

Hmm, yeah. So is some throttling by `threemaSafeBackupId` (for uploading?) already implemented?

**Comment by dbrgn** [comment]
Closing as out-of-scope. Throttling should be implemented in the reverse proxy (e.g. Nginx).
